ably/event_emitter.go

```diff
@@ -0,0 +1,176 @@
+package ably
+
+import (
+	"runtime/debug"
+	"sync"
+)
+
+type eventEmitter struct {
+	sync.Mutex
+	listeners listenersForEvent
+	logger    Logger
+}
+
+type emitterEvent = interface{}
+type emitterData = interface{}
+
+type listenersForEvent map[emitterEvent]listenerSet
+type listenerSet map[*eventListener]struct{}
+
+type eventListener struct {
+	handler func(emitterData)
```

Comment:
That's in the next PR :) I wanted to reuse eventEmitter for connection, channel and presence, so each of them will expose a type-safe API on top of a "generic" (untyped) one.

Reply:
Ok, but I still think there should be _something_ in the interface which tells the compiler (and developers working on the code) what kinds of things we expect to pass in here, for example:

```go
type emitterData interface {
	isEmitterData()
}

type ConnectionStateChange struct { }

func (ConnectionStateChange) isEmitterData() { }

type ChannelStateChange struct { }

func (ChannelStateChange) isEmitterData() { }
```

Reply:
But that doesn't really tell the compiler anything useful. This isn't a closed set of types that event emitter supports, it's really "any type". Doing that just makes the types declare that they're supposed to be used as events, but then we don't do anything useful with that information. It would be like having a generic linked list or something and requiring that element types implement a "isLinkedListElement" interface.

Sorry, I really don't see how to add more meaningful type-safety here.

Reply:
> that doesn't really tell the compiler anything useful

Since it's an internal interface, it tells the compiler (as well as developers and their IDEs) exactly what concrete types are expected to be passed as the first argument, this is absolutely useful information.

> This isn't a closed set of types that event emitter supports, it's really "any type"

Looking at the spec, there are exactly two types that implement `EventEmitter`:

1. A `RealtimeChannel` that emits `ChannelStateChange` events to its listeners:

```
class RealtimeChannel:
  embeds EventEmitter<ChannelEvent, ChannelStateChange?>
```

2. A `Connection` that emits `ConnectionStateChange` events to its listeners:

```
class Connection:
  embeds EventEmitter<ConnectionEvent, ConnectionStateChange>
```

If instead I chose to pass in, say, a `string`, then that really should fail to compile, because a `string`, as per the spec, isn't something that any listener expects.

> Doing that just makes the types declare that they're supposed to be used as events, but then we don't do anything useful with that information

We use that information to make sure that the code is spec compliant and only using those types as emitted events, no?

> It would be like having a generic linked list or something and requiring that element types implement a "isLinkedListElement" interface.

Well no, because a generic linked list typically isn't implementing a spec that declares what goes in the linked list. A better analogy would be a Go value that is expected to implement a [Protocol Buffer `oneof`](https://developers.google.com/protocol-buffers/docs/proto#using-oneof) (i.e. the PB stipulates that the value can be only one of a set of types), and what I'm proposing is exactly what [protoc-gen-go does](https://developers.google.com/protocol-buffers/docs/reference/go-generated#oneof).

Reply:
@lmars I do see your point. I just don't think there's much value in this:

> it tells the compiler (as well as developers and their IDEs) exactly what concrete types are expected to be passed as the first argument

While it unnecessarily ties event emitter to a closed set of types, which complicates testing.

But anyway, done at cc5c020.